src/server.js

```diff
@@ -29,19 +27,20 @@ const plugin =
           return next();
         }
         return new Promise((resolve, reject) => {
-          const oldEnd = ctx.res.end.bind(ctx.res);
-          // $FlowFixMe
-          ctx.res.end = (data, encoding, cb) => {
+          const {req, res} = ctx;
+          const listener = () => {
             ctx.respond = false;
-            return next()
-              .then(resolve)
-              .then(() => {
-                oldEnd(data, encoding, cb);
-              });
+            return next().then(resolve);
           };
-          handler(ctx.req, ctx.res, () => {
-            // $FlowFixMe
-            ctx.res.end = oldEnd;
+          res.on('end', listener);
```

Comment:
looking at how this is implemented in express, shouldn't we also listen on `finish`? asking because of https://github.com/jshttp/on-finished/blob/master/index.js#L104?

or maybe we could start using the `on-finished` package?